packages/x-date-pickers/src/MonthCalendar/MonthCalendar.tsx

```diff
@@ -54,14 +59,26 @@ const MonthCalendarRoot = styled('div', {
   name: 'MuiMonthCalendar',
   slot: 'Root',
   overridesResolver: (props, styles) => styles.root,
-})<{ ownerState: PickerOwnerState }>({
+  shouldForwardProp: (prop) => shouldForwardProp(prop) && prop !== 'monthsPerRow',
+})<{ ownerState: PickerOwnerState; monthsPerRow: 3 | 4 }>({
   display: 'flex',
   flexWrap: 'wrap',
-  alignContent: 'stretch',
-  padding: '0 4px',
+  justifyContent: 'space-evenly',
+  rowGap: 16,
+  padding: '8px 0',
   width: DIALOG_WIDTH,
   // avoid padding increasing width over defined
   boxSizing: 'border-box',
+  variants: [
+    {
+      props: { monthsPerRow: 3 },
+      style: { columnGap: 24 },
+    },
+    {
+      props: { monthsPerRow: 4 },
+      style: { columnGap: 0 },
+    },
+  ],
```

Comment:
I think the spacing makes sense here. If we want to have something similar to the previous version, we could do something like:
```suggestion
  justifyContent: 'space-between',
  rowGap: 16,
  width: DIALOG_WIDTH,
  // avoid padding increasing width over defined
  boxSizing: 'border-box',
  variants: [
    {
      props: { monthsPerRow: 3 },
      style: { padding: '8px 12px', },
    },
    {
      props: { monthsPerRow: 4 },
      style: { padding: '8px 2px', },
    },
  ],
```
That way the spacing between the columns is more dynamic and you shift the number of columns based on the horizontal padding. Tbh, I'm fine with either. One downside of both approaches is that if the `DIALOG_WIDTH` increases, so does the number of columns. 
We used `flexBasis` before and regardless of the parent's width, we had a fixed number of columns. We can't do that now because the button will stretch and look funky. We might want to mention this in the migration guide.